lib/puppet/type/file/data_sync.rb

```diff
@@ -42,7 +42,11 @@ def checksum_insync?(param, is, has_contents, &block)
           send resource[:loglevel], "[diff redacted]"
         else
           write_temporarily(param) do |path|
-            send resource[:loglevel], "\n" + diff(resource[:path], path)
+            diff_output = diff(resource[:path], path)
+            unless diff_output.force_encoding(Encoding::UTF_8).valid_encoding?
```

Comment:
@joshcooper , @gimmyxd , I added here a `force_encoding` to UTF-8 as by default in ruby 1.9.3 rspec string was ASCII-8BIT and always valid ...
I don't know is there could be other side effect of this change.

Reply:
I don't think force encoding is safe, as on Windows the default external encoding depends on the current code page, eg

```
irb(main):002:0> Encoding.default_external
=> #<Encoding:IBM437>
```

I'm also not sure how the diff program operates on non-ASCII content. For example, if you're using CP852, and you diff two UTF-16LE encoded files, `diff` should print the unicode character differences, but are those characters converted to CP852 when it writes to stdout? Or does it write the output as UTF-16LE. I assume it's the former?

I think the safest thing is only print the diff_output if `diff_output.encoding != Encoding::BINARY && diff_output.valid_encoding?`. Note binary and ascii-8bit are the same in ruby:

```
irb(main):003:0> Encoding::BINARY
=> #<Encoding:ASCII-8BIT>
```

Reply:
@joshcooper , I implemented the change. 
While Windows diff is probably not very usefull, I think the original issue is solved.
```
        --- C:/cygwin64/tmp/content_binary_file_test.O4vlo9	2019-11-25 12:39:34.967813100 +0000
        +++ C:/Users/Administrator/AppData/Local/Temp/puppet-file20191125-2468-34q1ks	2019-11-25 12:39:39.686842200 +0000
        @@ -1 +1 @@
        -╟╤ⁿä
        \ No newline at end of file
        +╟╤ⁿà
        \ No newline at end of file

```

Reply:
Maybe for the windows case, check if the text is ascii only? That may not be correct for non-EN_US locales though...

```
irb(main):002:0> "".ascii_only?
=> true
```

Reply:
Ok, the issue is the default Windows code page for our VMs is IBM 437, so the `diff` command interprets bytes `0xc7, 0xd1, 0xfc, 0x84` as characters in the IBM437 code page. Since those are valid encodings from bytes to chars, the diff program emits valid output, which results in a valid ruby string.

The same doesn't happen on Linux, because it typically uses UTF-8, and those bytes do not represent a valid UTF-8 encoded string.

```
irb(main):003:0> str = "\xc7\xd1\xfc\84"
=> "\xC7\xD1\xFC84"
irb(main):004:0> str.encoding
=> #<Encoding:UTF-8>
irb(main):005:0> str.valid_encoding?
=> false
```

Note if you change the windows code page to 65001 (aka UTF-8), then you'll get the same output as Linux:

```
C:\work\puppet> chcp 65001
Active code page: 65001

C:\work\puppet>bundle exec puppet apply manifest.pp --show_diff --diff "c:\Program Files\Git\usr\bin\diff.exe"
...
Notice: /Stage[main]/Main/File[c:\work\puppet\content.bin]/content:
Binary files c:/work/puppet/content.bin and C:/Users/ADMINI~1/AppData/Local/Temp/
puppet-file20191125-8132-1mkqagg differ
Notice: /Stage[main]/Main/File[c:\work\puppet\content.bin]/content: content changed '{md5}387d16f3cb61048e6fb16cfd81367ce8' to '{md5}c7930a40574e28aaaded83af90ed1
4a2'
Notice: Applied catalog in 9.84 seconds
```

However, I'm not a fan of changing the code page in the test, because that's not how people usually run puppet on Windows. Instead I'd update the test to expect whatever it is that puppet emits. The characters `\xc7\xd1\xfc\84` in IBM437 correspond to `\u255F\u2564\u207F\u00E4` in unicode, so you can use the latter in a literal ruby string.

```
irb(main):009:0> str = "\u255F\u2564\u207F\u00E4"
=> "╟╤ⁿä"
irb(main):010:0> str.encoding
=> #<Encoding:UTF-8>
irb(main):011:0> str.valid_encoding?
=> true
```

Reply:
Test expectations are now adapting to agent's `Encoding.default_external`